File: tests/correction/test_lenet.cpp
/**
 * @file test_lenet.cpp
 * @brief MNIST LeNet-5 卷积神经网络测试 - 支持多种激活函数
 * @version 1.0.0
 * @date 2026-06-04
 * @author Team Tech-Renaissance
 *
 * 设计原则：
 * - 基于经典LeNet-5架构，适配MNIST数据集
 * - 支持多种激活函数（ReLU/Tanh/SiLU等）
 * - 卷积层不支持bias（框架限制）
 * - 保持与mnist_best_adamw.cpp相同的测试框架
 *
 * 核心策略：
 * - 网络结构：Conv(6,5x5)→Pool→Conv(16,5x5)→Pool→Conv(120,5x5)→FC(84)→FC(10)
 * - 激活函数：可配置（默认tanh，经典LeNet-5使用）
 * - Bias：false（卷积层不支持bias）
 * - 优化器：AdamW（自适应学习率，稳定收敛）
 * - 学习率：CosineAnnealing+warmup(5)
 * - 数据增强：完整6种预处理链
 * - 训练轮数：100 epochs
 */

#include "renaissance.h"
#include <chrono>
#include <iomanip>
#include <iostream>
#include <string>
#include <cstdlib>

using namespace tr;

constexpr int kTotalEpochs = 100;

enum class TestMode { CPU, GPU, AMP };

const char* mode_name(TestMode m) {
    switch (m) {
    case TestMode::CPU: return "CPU  [FP32]";
    case TestMode::GPU: return "GPU  [FP32]";
    case TestMode::AMP: return "AMP  [FP16]";
    default:            return "???";
    }
}

struct CliConfig {
    TestMode mode = TestMode::GPU;
    std::string activation = "relu";  // 默认使用ReLU
};

CliConfig parse_cli(int argc, char** argv) {
    CliConfig cfg;
    bool mode_set = false;

    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        if (a == "--cpu") {
            if (mode_set) { std::cerr << "Multiple mode flags.\n"; std::exit(1); }
            cfg.mode = TestMode::CPU;
            mode_set = true;
        } else if (a == "--gpu") {
            if (mode_set) { std::cerr << "Multiple mode flags.\n"; std::exit(1); }
            cfg.mode = TestMode::GPU;
            mode_set = true;
        } else if (a == "--amp") {
            if (mode_set) { std::cerr << "Multiple mode flags.\n"; std::exit(1); }
            cfg.mode = TestMode::AMP;
            mode_set = true;
        } else if (a == "--activation" && i + 1 < argc) {
            cfg.activation = argv[++i];
        } else if (a == "--help") {
            std::cout << "Usage: " << argv[0] << " [OPTIONS]\n\n"
                      << "Mode flags (default: --gpu):\n"
                      << "  --cpu          Run on CPU\n"
                      << "  --gpu          Run on GPU (FP32)\n"
                      << "  --amp          Run on GPU (AMP FP16)\n\n"
                      << "Activation (default: tanh, classic LeNet-5):\n"
                      << "  --activation NAME   Choose from: relu, tanh, silu, relu6,\n"
                      << "                      leaky_relu, hardswish, elu, sigmoid\n\n"
                      << "Network Architecture:\n"
                      << "  Input: 1x32x32 (MNIST with padding)\n"
                      << "  C1: Conv(6, 5x5) + Activation + MaxPool(2x2)\n"
                      << "  C2: Conv(16, 5x5) + Activation + MaxPool(2x2)\n"
                      << "  C3: Conv(120, 5x5) + Activation\n"
                      << "  F4: FC(84) + Activation\n"
                      << "  F5: FC(10)\n\n"
                      << "Other:\n"
                      << "  --help         Show this message\n";
            std::exit(0);
        } else {
            std::cerr << "Unknown argument: " << a << "\n"
                      << "Use --help for usage information.\n";
            std::exit(1);
        }
    }
    return cfg;
}

Layer make_activation(const std::string& name) {
    if (name == "relu")       return relu();
    if (name == "tanh")       return tanh_act();
    if (name == "silu")       return silu();
    if (name == "relu6")      return relu6();
    if (name == "leaky_relu") return leaky_relu();
    if (name == "hardswish")  return hardswish();
    if (name == "elu")        return elu();
    if (name == "sigmoid")    return sigmoid();
    std::cerr << "Unknown activation: " << name << "\n"
              << "Supported: relu, tanh, silu, relu6, leaky_relu, hardswish, elu, sigmoid\n";
    std::exit(1);
}

int main(int argc, char** argv) {
    auto cfg = parse_cli(argc, argv);

    switch (cfg.mode) {
    case TestMode::CPU:
        GLOBAL_SETTING.use_cpu();
        break;
    case TestMode::GPU:
        GLOBAL_SETTING.use_gpu().amp(false);
        break;
    case TestMode::AMP:
        GLOBAL_SETTING.use_gpu().amp(true);
        break;
    }

    GLOBAL_SETTING
        .manual_seed(123)           // 固定随机种子，确保可复现
        .global_batch_size(128)     // 全局batch size，自动按world_size缩放
        .train_resolution(32)       // LeNet-5使用32x32输入
        .val_resolution(32)
        .use_tf32(true);           // 启用TF32加速

    PREPROCESSOR_SETTING
#ifdef _WIN32
        .dataset("mnist", "T:\\dataset\\mnist")
#else
        .dataset("mnist", "/root/epfs/dataset/mnist")
#endif
        .color_channels(1)
        .load_workers(8)           // 数据加载线程数
        .preprocess_workers(8)     // 预处理线程数
        .cpu_binding(true)        // CPU核心绑定
        .normalization(NormMode::MNIST)

        // 训练时使用前辈验证的最强预处理链
        .train_transforms(
            Pad(2),                      // 扩展到32x32，为后续操作提供空间
            RandomRotation(15.0f, 0),       // ±15度旋转
            RandomScale(0.9f, 1.1f),        // 0.9-1.1倍缩放
            RandomCrop(32),                 // 裁剪回32x32（LeNet-5输入尺寸）
            RandomAutocontrast(0.5f),       // 50%概率自动对比度调整
            RandomErasing(0.5f)             // 50%概率随机擦除
        )

        // 验证时不做增强
        .val_transforms(DoNothing())
        .commit();

    // LeNet-5架构：C1→C2→C3→F4→F5
    // 原始LeNet-5使用tanh，这里支持多种激活函数
    // 注意：卷积层不支持bias（框架限制）

    BluePrint lenet5 = seq(
        // C1: Conv(6, 5x5) + Activation + MaxPool(2x2)
        // 输入: 1x32x32 → 输出: 6x28x28 → 池化后: 6x14x14
        conv(6, 5, 1, 0),                  // out_channels=6, kernel=5, stride=1, pad=0
        make_activation(cfg.activation),   // 动态选择激活函数
        maxpool(2, 2, 0),                   // 2x2 最大池化

        // C2: Conv(16, 5x5) + Activation + MaxPool(2x2)
        // 输入: 6x14x14 → 输出: 16x10x10 → 池化后: 16x5x5
        conv(16, 5, 1, 0),                 // out_channels=16, kernel=5, stride=1, pad=0
        make_activation(cfg.activation),   // 动态选择激活函数
        maxpool(2, 2, 0),                   // 2x2 最大池化

        // C3: FC(120) — 5x5x16=400 输入，120 输出
        // 原 LeNet-5 C5 等价于全连接（kernel 覆盖整个特征图）
        fc(120, true),                     // 输入 400，输出 120
        make_activation(cfg.activation),   // 动态选择激活函数

        // F4: FC(84) + Activation
        fc(84, true),                      // 隐藏层，支持bias
        make_activation(cfg.activation),   // 动态选择激活函数

        // F5: FC(10) - 输出层
        fc(10, true)                        // 输出层，支持bias
    );

    DeepLearningTask task;
    task.model(lenet5)
        .loss(CrossEntropyLoss().label_smoothing(0.1f))  // Label Smoothing交叉熵损失

        // 初始化：Kaiming Uniform + FAN_IN（适配各种激活函数）
        .initializer(Initializer()
            .fc(InitKind::KAIMING_UNIFORM)
            .conv(InitKind::KAIMING_UNIFORM)
            .fan(FanMode::FAN_IN))

        .total_epochs(kTotalEpochs)           // 充分训练轮数

        // 优化器：AdamW（自适应学习率优化器）
        .optimizer(AdamW()
            .beta1(0.9f)                      // 标准动量参数
            .beta2(0.999f)                    // 标准二阶矩参数
            .eps(1e-8f)                       // 数值稳定性
            .weight_decay(1e-4f))             // 适中的权重衰减

        // 学习率调度：CosineAnnealing + warmup
        .scheduler(CosineAnnealingLR()
            .base_lr(0.001f)                  // AdamW标准初始学习率
            .eta_min(1e-6f)                   // 最小学习率
            .warmup(5)                        // 5轮预热
            .step_by_epoch())                 // 按epoch更新

        .validate_every(1, 1)                // 每轮验证
        .early_stop_by_top1(0.990f)          // LeNet-5目标：99%准确率
        .metrics(Metric::TRAIN_LOSS | Metric::VAL_LOSS | Metric::VAL_TOP1);

    std::cout << "\n=====================================\n"
              << "Renaissance MNIST LeNet-5 Test\n"
              << "=====================================\n"
              << " Mode: " << mode_name(cfg.mode) << "\n"
              << "=====================================\n"
              << "Network: LeNet-5\n"
              << "  C1: Conv(6,5x5) + " << cfg.activation << " + MaxPool(2x2)\n"
              << "  C2: Conv(16,5x5) + " << cfg.activation << " + MaxPool(2x2)\n"
              << "  C3: FC(120) + " << cfg.activation << "\n"
              << "  F4: FC(84) + " << cfg.activation << "\n"
              << "  F5: FC(10)\n"
              << "Activation: " << cfg.activation << "\n"
              << "Optimizer: AdamW (beta1=0.9, beta2=0.999, eps=1e-8, wd=1e-4)\n"
              << "Scheduler: CosineAnnealing + Warmup(5)\n"
              << "Augmentation: Pad+Rotation+Scale+Crop+Autocontrast+Erasing\n"
              << "Training: 100 epochs, batch_size=128\n"
              << "=====================================\n";

    task.compile();

    auto t0 = std::chrono::steady_clock::now();
    auto result = task.run();
    auto t1 = std::chrono::steady_clock::now();
    auto elapsed = std::chrono::duration<double>(t1 - t0).count();

    std::cout << std::fixed << std::setprecision(2)
              << "\n=====================================\n"
              << " Mode: " << mode_name(cfg.mode) << "\n"
              << "=====================================\n"
              << " Best Top-1: " << result.best_top1 * 100.0f << "%\n"
              << " Best Epoch: " << result.best_epoch << "\n"
              << " Total Time: " << elapsed << " s\n"
              << " Time per Epoch: " << elapsed / kTotalEpochs << " s\n"
              << "=====================================\n";

    // 根据准确率给出评价（LeNet-5标准稍低于MLP）
    if (result.best_top1 >= 0.990f) {
        std::cout << "EXCELLENT! Accuracy >= 99.0%\n";
        std::cout << "Target achieved: LeNet-5 goal met!\n";
    } else if (result.best_top1 >= 0.985f) {
        std::cout << "VERY GOOD! Accuracy >= 98.5%\n";
        std::cout << "Performance: Above LeNet-5 baseline\n";
    } else if (result.best_top1 >= 0.980f) {
        std::cout << "GOOD. Accuracy >= 98.0%\n";
        std::cout << "Performance: Acceptable LeNet-5 accuracy\n";
    } else if (result.best_top1 >= 0.975f) {
        std::cout << "ACCEPTABLE. Accuracy >= 97.5%\n";
        std::cout << "Consider: Increasing epochs or adjusting hyperparameters\n";
    } else {
        std::cout << "NEEDS IMPROVEMENT. Accuracy < 97.5%\n";
        std::cout << "Recommendations:\n";
        std::cout << "  1. Increase training epochs to 150+\n";
        std::cout << "  2. Try different activation functions (relu, silu)\n";
        std::cout << "  3. Adjust weight_decay (try 5e-4 or 5e-5)\n";
        std::cout << "  4. Increase network capacity (more channels)\n";
    }

    std::cout << "=====================================\n";

    // 返回值：99.0%以上返回0（成功），否则返回1（失败）
    return result.best_top1 >= 0.990f ? 0 : 1;
}

File: tests/bn/test_lenet.py
import argparse
import math
import time

import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader
from torchvision import datasets, transforms

# ---------------------------------------------------------------------------
# RandomAutocontrast compatibility (torchvision >= 0.10)
# ---------------------------------------------------------------------------
try:
    _RandomAutocontrast = transforms.RandomAutocontrast
    _HAS_AUTOCONTRAST = True
except AttributeError:
    _HAS_AUTOCONTRAST = False

parser = argparse.ArgumentParser(description="PyTorch MNIST LeNet-5 benchmark")
parser.add_argument("--cpu", action="store_true", help="Run on CPU (FP32)")
parser.add_argument("--gpu", action="store_true", help="Run on GPU (FP32)")
parser.add_argument("--amp", action="store_true", help="Run on GPU with AMP (FP16)")
parser.add_argument("--activation", default="tanh",
                    help="Activation function (relu, tanh)")


class LeNet5(nn.Module):
    def __init__(self, activation="tanh"):
        super().__init__()

        # 选择激活函数：仅支持 ReLU 和 Tanh
        if activation == "relu":
            act = nn.ReLU()
        elif activation == "tanh":
            act = nn.Tanh()
        else:
            raise ValueError(f"Unknown activation: {activation}. Supported: relu, tanh")

        # 网络结构（与 Renaissance test_lenet.cpp 对齐）：
        # channel_padding (implicit in NCHW) -> conv(8,5,1,2) -> act -> avgpool(2,2,0)
        #   -> conv(16,5,1,0) -> act -> avgpool(2,2,0)
        #   -> fc(120) -> act -> fc(88) -> act -> fc(10)
        self.features = nn.Sequential(
            # C1: 8通道 5x5 卷积, padding=2 -> 28x28x8
            nn.Conv2d(1, 8, kernel_size=5, stride=1, padding=2, bias=False),
            act,
            # S2: 2x2 AvgPool, stride=2 -> 14x14x8
            nn.AvgPool2d(kernel_size=2, stride=2, padding=0),
            # C3: 16通道 5x5 卷积, padding=0 -> 10x10x16
            nn.Conv2d(8, 16, kernel_size=5, stride=1, padding=0, bias=False),
            act,
            # S4: 2x2 AvgPool, stride=2 -> 5x5x16
            nn.AvgPool2d(kernel_size=2, stride=2, padding=0),
        )

        self.classifier = nn.Sequential(
            # Flatten 5x5x16 = 400
            nn.Flatten(),
            # fc(120), 有bias
            nn.Linear(400, 120, bias=True),
            act,
            # fc(88), 有bias（8的倍数，兼容AMP模式）
            nn.Linear(120, 88, bias=True),
            act,
            # 输出: 10单元, 有bias
            nn.Linear(88, 10, bias=True)
        )

        self._init_weights()

    def _init_weights(self):
        """Match TR4: Kaiming Uniform (FAN_IN) for all linear/conv layers"""
        for m in self.modules():
            if isinstance(m, (nn.Linear, nn.Conv2d)):
                nn.init.kaiming_uniform_(m.weight, a=0, mode='fan_in')
                # Match TR4: bias is initialized to ZEROS
                if m.bias is not None:
                    nn.init.zeros_(m.bias)

    def forward(self, x):
        x = self.features(x)
        x = self.classifier(x)
        return x


WARMUP_EPOCHS = 5
TOTAL_EPOCHS  = 100
ETA_MIN       = 1e-6
BASE_LR       = 0.001


def warmup_cosine_lambda(epoch):
    """
    Match TR4 CosineAnnealingLR + warmup(5) behaviour exactly.
    Epoch 1 (epoch=0): lr = 0
    Epoch 2 (epoch=1): lr = 0.0002
    ...
    Epoch 6 (epoch=5): lr = 0.001  (base_lr reached)
    Epoch 7+          : cosine annealing down to eta_min
    """
    if epoch < WARMUP_EPOCHS:
        return epoch / WARMUP_EPOCHS          # 0 -> 0.0, 1 -> 0.2, ..., 4 -> 0.8
    progress = (epoch - WARMUP_EPOCHS) / (TOTAL_EPOCHS - WARMUP_EPOCHS)
    cos_val = 0.5 * (1.0 + math.cos(math.pi * progress))
    return (ETA_MIN + (BASE_LR - ETA_MIN) * cos_val) / BASE_LR


if __name__ == '__main__':
    args = parser.parse_args()

    if args.amp:
        mode = "AMP"
        device = torch.device("cuda")
        use_amp = True
    elif args.gpu:
        mode = "GPU"
        device = torch.device("cuda")
        use_amp = False
    else:
        mode = "CPU"
        device = torch.device("cpu")
        use_amp = False

    print("===========================================", flush=True)
    print(" PyTorch MNIST LeNet-5 Benchmark", flush=True)
    print("===========================================", flush=True)
    print(f" Mode:       {mode}", flush=True)
    print(f" Device:     {device}", flush=True)
    print(f" Network:    LeNet-5 (2 conv + 3 fc)", flush=True)
    print(f"   C1: conv(8,5,1,2) -> activation -> avgpool(2,2,0)", flush=True)
    print(f"   C3: conv(16,5,1,0) -> activation -> avgpool(2,2,0)", flush=True)
    print(f"   Flatten(400) -> fc(120) -> activation -> fc(88) -> activation -> fc(10)", flush=True)
    print(f" Conv bias:  false", flush=True)
    print(f" FC bias:    true", flush=True)
    print(f" Activation: {args.activation}", flush=True)
    print(f" Optimizer:  AdamW (beta1=0.9, beta2=0.999, eps=1e-8, wd=1e-4)", flush=True)
    print(f" Scheduler:  CosineAnnealing + Warmup(5)", flush=True)
    if not _HAS_AUTOCONTRAST:
        print(" Warning:    RandomAutocontrast not available (torchvision < 0.10)", flush=True)
    print("===========================================", flush=True)

    if device.type == "cuda":
        torch.backends.cuda.matmul.allow_tf32 = True
        torch.backends.cudnn.allow_tf32 = True
        torch.set_float32_matmul_precision('high')

    torch.manual_seed(123)
    batch_size = 128
    epochs = 100

    # -----------------------------------------------------------------------
    # Transforms: match TR4 as closely as possible
    # TR4 order: Pad -> RandomRotation -> RandomScale -> RandomCrop
    #            -> RandomAutocontrast -> (ToTensor + Normalize + RandomErasing)
    # -----------------------------------------------------------------------
    train_transform_list = [
        transforms.Pad((2, 2, 2, 2), fill=0),
        transforms.RandomRotation(15, fill=0),
        transforms.RandomAffine(0, scale=(0.9, 1.1)),
        transforms.RandomCrop(28),
    ]
    if _HAS_AUTOCONTRAST:
        train_transform_list.append(transforms.RandomAutocontrast(p=0.5))
    train_transform_list += [
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,)),
        transforms.RandomErasing(p=0.5),
    ]
    train_transform = transforms.Compose(train_transform_list)

    val_transform = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,)),
    ])

    train_set = datasets.MNIST("T:/dataset/mnist", train=True,  download=False, transform=train_transform)
    val_set   = datasets.MNIST("T:/dataset/mnist", train=False, download=False, transform=val_transform)

    pin_mem = (device.type == "cuda")
    train_loader = DataLoader(train_set, batch_size=batch_size, shuffle=True,
                              pin_memory=pin_mem, num_workers=8, persistent_workers=True)
    val_loader   = DataLoader(val_set,   batch_size=batch_size, shuffle=False,
                              pin_memory=pin_mem, num_workers=8, persistent_workers=True)

    model = LeNet5(activation=args.activation).to(device)

    criterion = nn.CrossEntropyLoss(label_smoothing=0.1)

    # 参数分组：weight 用 AdamW (wd=1e-4)，bias/1D 参数用 Adam (wd=0.0)
    # 匹配 TR4 行为：BN Weight、BN Bias、FC Bias 固定 WD=0
    weight_params = []
    bias_params = []
    for name, param in model.named_parameters():
        if not param.requires_grad:
            continue
        if 'bias' in name or len(param.shape) == 1:
            bias_params.append(param)
        else:
            weight_params.append(param)

    param_groups = [
        {'params': weight_params, 'weight_decay': 1e-4},
        {'params': bias_params, 'weight_decay': 0.0}
    ]
    optimizer = optim.AdamW(param_groups, lr=BASE_LR,
                            betas=(0.9, 0.999), eps=1e-8)
    scheduler = optim.lr_scheduler.LambdaLR(optimizer, lr_lambda=warmup_cosine_lambda)

    scaler = torch.amp.GradScaler("cuda") if use_amp else None

    best_acc = 0.0
    best_epoch = 0
    t0 = time.perf_counter()

    for epoch in range(epochs):
        ep_t0 = time.perf_counter()

        model.train()
        train_loss = 0.0

        for data, target in train_loader:
            data, target = data.to(device), target.to(device)
            optimizer.zero_grad()

            if use_amp:
                with torch.amp.autocast("cuda"):
                    output = model(data)
                    loss = criterion(output, target)
                scaler.scale(loss).backward()
                scaler.step(optimizer)
                scaler.update()
            else:
                output = model(data)
                loss = criterion(output, target)
                loss.backward()
                optimizer.step()

            train_loss += loss.item()

        avg_train_loss = train_loss / len(train_loader)

        model.eval()
        val_loss = 0.0
        correct = 0
        total = 0

        with torch.no_grad():
            for data, target in val_loader:
                data, target = data.to(device), target.to(device)

                if use_amp:
                    with torch.amp.autocast("cuda"):
                        output = model(data)
                        val_loss += criterion(output, target).item()
                else:
                    output = model(data)
                    val_loss += criterion(output, target).item()

                pred = output.argmax(dim=1)
                correct += pred.eq(target).sum().item()
                total += target.size(0)

        avg_val_loss = val_loss / len(val_loader)
        acc = 100.0 * correct / total

        if acc > best_acc:
            best_acc = acc
            best_epoch = epoch + 1

        current_lr = optimizer.param_groups[0]['lr']
        ep_t1 = time.perf_counter()

        print(f"Epoch {epoch+1:3d}: train_loss={avg_train_loss:.6f}  val_loss={avg_val_loss:.6f}  "
              f"val_top1={acc:.2f}%  lr={current_lr:.6f}  time={ep_t1 - ep_t0:.1f}s", flush=True)

        scheduler.step()

    t1 = time.perf_counter()
    print("\n===========================================", flush=True)
    print(f" Mode:       {mode}", flush=True)
    print(f" Best Top-1: {best_acc:.2f}%", flush=True)
    print(f" Best Epoch: {best_epoch}", flush=True)
    print(f" Total Time: {t1 - t0:.2f} s", flush=True)
    print(f" Time/Epoch: {(t1 - t0) / epochs:.2f} s", flush=True)
    print("===========================================", flush=True)

    # 打印评估结果
    if best_acc >= 98.5:
        print("EXCELLENT! Accuracy >= 98.5%", flush=True)
        print("Target achieved: Goal met!", flush=True)
    elif best_acc >= 98.0:
        print("GOOD! Accuracy >= 98.0%", flush=True)
        print("Performance: Above baseline", flush=True)
    elif best_acc >= 97.5:
        print("ACCEPTABLE. Accuracy >= 97.5%", flush=True)
        print("Consider: Increasing epochs or adjusting hyperparameters", flush=True)
    else:
        print("NEEDS IMPROVEMENT. Accuracy < 97.5%", flush=True)
        print("Recommendations:", flush=True)
        print("  1. Increase training epochs to 150+", flush=True)
        print("  2. Adjust weight_decay (try 5e-4 or 5e-5)", flush=True)
        print("  3. Increase network width to fc(1152, true)", flush=True)
    print("===========================================", flush=True)
